src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -666,14 +673,65 @@ void AtlasEngine::_flushBufferLine()
     // This would seriously blow us up otherwise.
     Expects(_api.bufferLineColumn.size() == _api.bufferLine.size() + 1);
 
+    const auto beg = _api.bufferLine.data();
+    const auto len = _api.bufferLine.size();
+    size_t segmentBeg = 0;
+    size_t segmentEnd = 0;
+    bool custom = false;
+
+    if (!_hackWantsBuiltinGlyphs)
+    {
+        _mapRegularText(0, len);
+        return;
+    }
+
+    while (segmentBeg < len)
+    {
+        segmentEnd = segmentBeg;
+        do
+        {
+            auto i = segmentEnd;
+            char32_t codepoint = beg[i++];
+            if (til::is_leading_surrogate(codepoint) && i < len)
+            {
+                codepoint = til::combine_surrogates(codepoint, beg[i++]);
+            }
+
+            const auto c = BuiltinGlyphs::IsCustomGlyph(codepoint) || BuiltinGlyphs::IsSoftFontChar(codepoint);
+            if (custom != c)
+            {
+                break;
+            }
+
+            segmentEnd = i;
+        } while (segmentEnd < len);
+
+        if (segmentBeg != segmentEnd)
+        {
+            if (custom)
+            {
+                _mapBuiltinGlyphs(segmentBeg, segmentEnd);
+            }
+            else
+            {
+                _mapRegularText(segmentBeg, segmentEnd);
+            }
+        }
+
+        segmentBeg = segmentEnd;
+        custom = !custom;

```

Comment:
this isn't a request for this PR, but...

we use this paradigm *a lot*.

* batch up data until point x
* do thing x1
* batch up data until point y
* do thing y1
* repeat

VT parsing, segmentation, splitting up a run of text to make some of it bold (like in command palette search results), etc etc etc

Is there some way to model this as a til helper? `til::batching_split_multistate` or something?

Reply:
I started writing a function for this but realized that the code kinda sucks a little for this particular use case. Here, we need to iterate over code points while retaining iterators in code units, which is quite annoying to do with C++'s weird iterators. `ranges` would probably simplify this, but alas they're extremely expensive. I do think we should have such a function as well, but I'm not sure if this here would be a good place to use it.